### ISSUE TITLE
MAME (and MAME64) hit zip 4GB limitation

### ISSUE THREAD

**Issue by NULUSIOS** [mamedev/mame#785]
MAME supports zipped resource folders (like snapshots), now with embedded MEWUI, far more - but as we discovered here...
( link removed )
...hits zip library's 4GB limit.

As MAME and emulated computing goes further in the 21st century, those resources will only grow bigger.

Probably needs a zip library able to handle >4GB archives, at least for the x64 MAME version.

**Comment by cuavas** [comment]
You can use 7zip archives with MAME as long as no individual file in the archive exceeds 4GB.

**Comment by LnmVolbo** [comment]
Perhaps there is a misunderstanding.  None of the individual files exceeds 4GB.  The whole archive exceeds 4GB because there are enough smaller files in it.

When a .zip file grows beyond 4GB, a pointer in the central directory structure needs more than 4 bytes to "point" to the offset to where the data starts for each file.  When this happens, the pointer is set to 0xFFFFFFFF, and another field is used, called the "extra field" to hold 8 bytes for the offset location pointer.
 
Unsophisticated .zip libraries are ignorant of this and fail to retrieve any files from the archive that use the "extra field" for the offset location pointer.

Not insisting that this be looked at, just trying to explain in more detail what the issue is.

-----

Some editorial content follows.

Since MAME is expressing this (feature of ignoring files from zip archives over 4GB), one can conclude it is not a fully compliant 64-bit application.  Not ironic that legacy (such as WinXP and/or others) support is dropped, yet the zip subsystem keeps one foot firmly planted int he 32 bit world?

Of course MAME doesn't need to handle this (right?).  There are already some front ends that can handle this issue (because they utilize zip libraries that work with 4 GB+ archives).  What was the point of absorbing MEWUI into the collective when users chose another front end to handle this issue?

And what's the point of support reading these 64-bit zips?  There is never going to be a ROM over 4 GB, right?  Didn't Haze recently construct a special CHD for a 8GB encryption table for that snowboarding game back in January?  Now that was a single big file over 4GB.  Maybe that recollection is confusing the issue.

Of what possible use could a fully compliant 64-bit zip subsystem be to MAME in the future?  No one can say, if there isn't one, then no one will know.  Perhaps it is a poor utilization of resources to add it now.  Not adding it will not cripple the MAME project. But adding it will make MAME a tiny bit more flexible.

**Comment by rb6502** [comment]
I'm reopening this because there's a trivial fix: upgrading our version of Info-Zip's unzip library to the 6.0 version which handles ZIP64 extensions (and contains a number of security and correctness fixes).

Given that this issue shows at least 3 people have strong opinions about this, I expect a pull request from someone that does this shortly.  Don't disappoint me.

**Comment by cuavas** [comment]
We don't use the info zip library for reading zip files -MAME has its own code for walking the ecd and calls zlib to decompress the data. The relevant code is in src/lib/util/unzip.cpp and hidden entirely behind a C++ interface class. It should be easy to implement zip64 without high impact on the restb of the code.

**Comment by rb6502** [comment]
We reinvent one of the most commonly sourced wheels in applied computer science?  Alright then.  I still expect a pull request, but the difficulty's gone up a hair.

**Comment by LnmVolbo** [comment]
Thanks for reconsidering.  I don't have a pull request ready right now.  I'll see what I can come up with.  If anyone already has something in mind, please feel free to submit it.

**Comment by LnmVolbo** [comment]
I found the code that **cuavas** pointed out, thanks.  It seems to me that the correct routine to modify is "get_compressed_data_offset" at src/lib/util/unzip.cpp, starting at line 607.

I've come up with some pseudo code.  If anyone is so motivated, please provide critique.

```c++
//	make a new 8 byte pointer, set it to the current 4 byte pointer "m_header.local_header_offset"
//	test to see if the 4 byte pointer is 0xffffffff and that the extra size field length is not zero
//		allocate a temporary buffer for the extra field and load it
//		check the "header" of the extra field to see that it has one 8 byte entry
//			load the 8 byte pointer with the 8 byte payload of the extra field
//		dispose of the temporary buffer
//	continue the execution of the existing routine with the new 8-byte pointer.

**Comment by LnmVolbo** [comment]
Making progress, I've got some code written that compiles w/o errors, just needed a 6GB file to test it on...

Hmm... tried the revised MAME with the new code, but it doesn't seem to work.  By that I mean it does work (the gui displays .png files from the .zip file) for 4GB- files but not 4GB+ files.

If anyone wants to see the code: https://gist.github.com/LnmVolbo/eeea541d8ec0266d25e97f79d0a6eb8f

I found a way (I think) to extend the existing data structure to avoid creating a temporary buffer as I outlined in the above pseudocode.  Perhaps my assumptions are erroneous, since it doesn't work.

Unfortunately I am not well acquainted with MAME's debugger, so progress may be slow from here on.  If any interested parties have any ideas, I'm all ears.

Be forewarned, the code is experimental and has not been tested extensively.

**Comment by LnmVolbo** [comment]
I just noticed that the routine that finds the central directory uses a 4 byte offset to search.  I will amend the code to use an 8 byte offset and report back.

**Comment by ajrhacker** [comment]
Volbo, it seems that a number of substructure fields in the code you linked still need to be expanded from 32 bits to 64 bits and initialized accordingly. Otherwise it does no good to return a 64-bit value from current_uncompressed_length, for instance.

**Comment by LnmVolbo** [comment]
ajrhacker, thanks for the input.

Well, to get started with this, I'm not trying to support **individual** archive content files that are 4GB+, just an **entire** archive length of 4GB+.  I am not attempting to support the fields _compressed_length_ or _uncompressed_length_ at this time.  Once the basic functionality is established, it can be expanded.  There are all sorts of things that can go into the extra field.  That's why I am screening for the header with the id of "1" and size "8".  Because the 4GB+ .zip file I am working with has only that configuration in the archived entries beyond the point where the _local_header_offset_ overflows past 0x0ffffffff.

Or are you talking about some other fields I have overlooked? if so, please elaborate.

By the way, it is possible to place comments right on the gist page (scroll down to the bottom to see it).  This avoids clutter on the master announce page.

**Comment by LnmVolbo** [comment]
Sorry, I am not familiar enough with github to tell if my pull request registered properly.

Any assistance is appreciated.

**Comment by LnmVolbo** [comment]
Another stab at this.

I feel bad about wasting your time with my previous pull requests.

Hopefully this one is more worthy?

https://github.com/mamedev/mame/pull/793/files

**Comment by cuavas** [comment]
Implemented in f5f9b45da8bdd46939e64ebcec1a6afc13063656
